File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let f=0,l=nums.length-1;
    while(f<=l){
        let m=Math.floor(f+(l-f)/2);
        if(nums[m]===target)return m;
        if(nums[f]<=nums[m]){
            if(nums[f]<=target && target<=nums[m]){
                l=m;
            }else f=m+1;
        }else{
            if(nums[m+1]<=target && target<=nums[l]){
                f=m+1;
            }else l=m;
        }
    }
    return -1;
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int f=0,l=nums.size()-1;
        while(f<=l){
            int mid=f+(l-f)/2;
            if(nums[mid]==target) return mid;
            if(nums[f]<=nums[mid]){
                if(nums[f]<=target && target<=nums[mid]){
                    l=mid-1;
                }else f=mid+1;
            }else{
                if(nums[mid]<=target && target<=nums[l]){
                    f=mid+1;
                }else l=mid-1;
            }
        }
        return -1;
    }
};